src/ws.nim

```diff
@@ -0,0 +1,570 @@
+import chronos, chronicles, httputils, strutils, base64, std/sha1, random,
+    streams, nativesockets, uri, times, chronos.timer, tables
+
+const
+  MaxHttpHeadersSize = 8192        # maximum size of HTTP headers in octets
+  MaxHttpRequestSize = 128 * 1024  # maximum size of HTTP body in octets
+  HttpHeadersTimeout = timer.seconds(120) # timeout for receiving headers (120 sec)
+  HttpBodyTimeout = timer.seconds(12)     # timeout for receiving body (12 sec)
+  HeadersMark = @[byte(0x0D), byte(0x0A), byte(0x0D), byte(0x0A)]
+  WebsocketUserAgent* = "nim-ws (https://github.com/status-im/nim-ws)"
+
+type
+  ReadyState* = enum
+    Connecting = 0 # The connection is not yet open.
+    Open = 1       # The connection is open and ready to communicate.
+    Closing = 2    # The connection is in the process of closing.
+    Closed = 3     # The connection is closed or couldn't be opened.
+
+  WebSocket* = ref object
+    tcpSocket*: StreamTransport
+    version*: int
+    key*: string
+    protocol*: string
+    readyState*: ReadyState
+    masked*: bool # send masked packets
+    header*: HttpHeaders
+
+  AsyncCallback = proc (transp: StreamTransport,
+      header: HttpRequestHeader): Future[void] {.closure, gcsafe.}
+  HttpServer* = ref object of StreamServer
+    callback: AsyncCallback
+
+  ReqStatus = enum
+    Success, Error, ErrorFailure
+
+  WebSocketError* = object of IOError
+
+  HttpHeaders* = ref object
+    table*: TableRef[string, seq[string]]
```

Comment:
just `Table` - `ref` loses mutability protections in the language and is susceptible to `nil` issues.